src/contexts/SetComponents/SetComponentsProvider.tsx

```diff
@@ -0,0 +1,146 @@
+        })
+        Promise.all(dataPositions).then(sortPositionsByPercentOfSet).then(setDataComponents)
+      })
+    }
+  }, [ethereum, tokenList, dpiPrice, mviPrice, bedPrice, eth2xfliPrice, btc2xfliPrice, dataPrice])
+
+  return (
+    <SetComponentsContext.Provider
+      value={{
+        dpiComponents: dpiComponents,
+        mviComponents: mviComponents,
+        bedComponents: bedComponents,
+        eth2xfliComponents: eth2xfliComponents,
+        btc2xfliComponents: btc2xfliComponents,
+        dataComponents: dataComponents,
+      }}
+    >
+      {children}
+    </SetComponentsContext.Provider>
+  )
+}
+
+async function convertPositionToSetComponent(position: Position, tokenList: Token[], componentPriceUsd: number, setPriceUsd: number): Promise<SetComponent> {
+  const token = getTokenForPosition(tokenList, position)
+  const quantity = new BigNumber(position.unit.toString()).div(new BigNumber(10).pow(getDecimalPlaces(token.symbol)));
+  const totalPriceUsd = quantity.multipliedBy(componentPriceUsd);
+  const percentOfSet = totalPriceUsd.dividedBy(setPriceUsd).multipliedBy(100)
+
+  if (token.symbol.toLowerCase() === "ceth" || token.symbol.toLowerCase() === 'usdc') {
+    // console.log(`componentPriceUsd`, componentPriceUsd)
+    // console.log(`quantity`, quantity.toString())
+    console.log(`${token.symbol.toLowerCase()} position.unit`, position.unit.toString())
+    console.log(`quantity`, quantity.toString())
+  }
+
+  return {
+    address: position.component,
+    id: token.name.toLowerCase(),
+    quantity: quantity.toString(),
+    symbol: token.symbol,
+    name: token.name,
+    image: token.logoURI,
+    totalPriceUsd: totalPriceUsd.toString(),
+    percentOfSet: percentOfSet.toPrecision(3).toString(),
+    percentOfSetNumber: percentOfSet,
+  }
+}
+
+function getTokenForPosition(tokenList: Token[], position: Position): Token {
+  const matchingTokens = tokenList.filter(t => t.address.toLowerCase() === position.component.toLowerCase())
+  if (matchingTokens.length === 0) {
+    console.warn(`No token for position ${position.component} exists in token lists`)
+  } else if (matchingTokens.length > 1) {
+    console.warn(`Multiple tokens for position ${position.component} exist in token lists`)
+  }
+  return matchingTokens[0]
+}
+
+function sortPositionsByPercentOfSet(components: SetComponent[]): SetComponent[] {
+  return components.sort((a, b) => b.percentOfSetNumber.comparedTo(a.percentOfSetNumber))
+}
+
+function getPositionPrices(setDetails: SetDetails): Promise<any> {
+  const componentAddresses = setDetails.positions.map(p => p.component)
+  return fetch(`https://api.coingecko.com/api/v3/simple/token_price/${ASSET_PLATFORM}?vs_currencies=${VS_CURRENCY}&contract_addresses=${componentAddresses}`)
+    .then(response => response.json())
+    .catch(e => console.error(e))
+}
+
+// HACKHACK is there a safer way to fetch decimal places for these components?
```

Comment:
Do reviewers know of a safer way to determine this? I naively assumed all component quantities would have 18 significant figures but I had to divide by 4, 6, 8, or 18 to get the component quantities for products like Eth2xFli to match up with what is visible on https://www.tokensets.com/explore

Reply:
That's actually a read only function on the contract called `decimals()` with no params, returns a BN. 

This will need to be checked for sure. MVI has a couple of tokens with different significant digits (WHALE comes to mind)

Reply:
yep WHALE and WAXE are definitely oddities so I special cased them in this function. Thanks for sharing the `decimals()` function. I manually verified the tokens that I special cased in the function and they all line up.

Symbol | decimals | contract
--- | --- | ---
WHALE | 4 | https://etherscan.io/address/0x9355372396e3F6daF13359B7b607a3374cc638e0#readContract
USDC | 6 | https://etherscan.io/address/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48#readProxyContract
WBTC | 8 | https://etherscan.io/address/0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599#readContract
WAXE | 8 | https://etherscan.io/address/0x7a2Bc711E19ba6aff6cE8246C546E8c4B4944DFD#readProxyContract
CETH | 8 | https://etherscan.io/address/0x4Ddc2D193948926D02f9B1fE9e1daa0718270ED5#readContract
CWBTC | 8 | https://etherscan.io/token/0xC11b1268C1A384e55C48c2391d8d480264A3A7F4#readContract

Considering this PR diff is huge, I'm inclined to merge this as-is and tackle reading `decimals()` from on-chain for all token contracts in a fast-follow PR if that's okay with reviewers.

Reply:
Created https://github.com/SetProtocol/index-ui/issues/499

Reply:
That makes this _way_ easier than I expected. Will fix in this PR